### PR TITLE
indexserver: add IndexConcurrency for concurrent index jobs

### PR DESCRIPTION
**Pull request by keegancsmith**
This is a tunable which we only expect to be used when bootstrapping a
large number of repos to be indexed. The previous commit introduced some
concurrency control to prevent processing the same repository
concurrently. This allows us to naively spin up N goroutines to process
the queue. Note: the queue already contains concurrency control.

Now what is needed is some better observability around what is currently
being processed / etc. The net/trace instrumentation does expose this,
but we need something smoother if we polish up this feature.

Meta: The amount of copy paste to add a new variable is a bit of a
smell, but we can address that in a later commit.

Test Plan: go test. Manually start up a fresh instance with 10s of repos
to index and observe concurrent indexing.

plz-review-url: https://plz.review/review/6342

### CODE REVIEW

File: cmd/zoekt-sourcegraph-indexserver/main.go [hunk 1]
```diff
@@ -364,6 +367,15 @@ func (s *Server) Run() {
 	}()
 
 	// In the current goroutine process the queue forever.
```
Comment (mpimenov) [inline cmd/zoekt-sourcegraph-indexserver/main.go]: The comment is outdated now?

File: cmd/zoekt-sourcegraph-indexserver/main.go [hunk 2]
```diff
@@ -969,6 +982,7 @@ func (rc *rootConfig) registerRootFlags(fs *flag.FlagSet) {
 	fs.DurationVar(&rc.mergeInterval, "merge_interval", time.Hour, "run merge this often")
 	fs.Int64Var(&rc.targetSize, "merge_target_size", getEnvWithDefaultInt64("SRC_TARGET_SIZE", 2000), "the target size of compound shards in MiB")
 	fs.Int64Var(&rc.minSize, "merge_min_size", getEnvWithDefaultInt64("SRC_MIN_SIZE", 1800), "the minimum size of a compound shard in MiB")
+	fs.Int64Var(&rc.indexConcurrency, "index_concurrency", getEnvWithDefaultInt64("SRC_INDEX_CONCURRENCY", 1), "the number of concurrent index jobs to run.")
```
Comment (stefanhengl) [inline cmd/zoekt-sourcegraph-indexserver/main.go:984]: I guess we could use fs.IntVar instead and avoid the conversion later?

**Review by stefanhengl — APPROVED** [review]
Nice!

File: cmd/zoekt-sourcegraph-indexserver/main.go (hunk 1 above)
Comment (keegancsmith) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: indeed, will remove.

File: cmd/zoekt-sourcegraph-indexserver/main.go (hunk 2 above)
Comment (keegancsmith) [reply cmd/zoekt-sourcegraph-indexserver/main.go:984]: I did this to get the free helper `getEnvWithDefaultInt64`. To be honest, this all needs a bit of a cleanup. Want me to tackle now or later?

**Review by keegancsmith — COMMENTED** [review]
1 comment(s) resolved

**Review by mpimenov — APPROVED** [review]

File: cmd/zoekt-sourcegraph-indexserver/main.go (hunk 2 above)
Comment (stefanhengl) [reply cmd/zoekt-sourcegraph-indexserver/main.go:984]: later is fine ;-)

**Review by keegancsmith — COMMENTED** [review]
Updated test plan to include command run to test. Looks good so will proceed with merge. 

 
 
 2 comment(s) resolved

**Pull request merged**